// @discloudapp/api-types/payload/v2/DiscloudConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.discloudConfigScopes = exports.discloudConfigRequiredScopes = exports.DiscloudConfigScopes = exports.APTPackages = exports.APT = void 0;
exports.APT = {
    canvas: [
        "fonts-liberation",
        "libcairo2-dev",
        "libgbm-dev",
        "libgif-dev",
        "libjpeg-dev",
        "libpango1.0-dev",
        "librsvg2-dev",
    ],
    ffmpeg: [
        "ffmpeg",
    ],
    java: [
        "default-jre",
    ],
    libgl: [
        "cmake",
        "libsm6",
        "libxext6",
        "libtool",
        "make",
    ],
    mysql: [
        "default-mysql-client",
    ],
    openssl: [
        "pkg-config",
        "libssl-dev",
    ],
    puppeteer: [
        "libasound2",
        "libatk1.0-0",
        "libatk-bridge2.0-0",
        "libcairo2",
        "libcups2",
        "libdrm-dev",
        "libdrm2",
        "libgbm-dev",
        "libgbm1",
        "libglib2.0-0",
        "libgtk-3-0",
        "libnss3",
        "libpango-1.0-0",
        "libx11-xcb-dev",
        "libxcomposite1",
        "libxdamage1",
        "libxfixes3",
        "libxkbcommon0",
        "libxrandr2",
        "libxshmfence-dev",
    ],
    tesseract: [
        "tesseract-ocr",
    ],
    tools: [
        "curl",
        "git",
        "openssh-client",
        "wget",
    ],
};
exports.APTPackages = Object.keys(exports.APT);
var DiscloudConfigScopes;
(function (DiscloudConfigScopes) {
    DiscloudConfigScopes["APT"] = "APT";
    DiscloudConfigScopes["AUTORESTART"] = "AUTORESTART";
    DiscloudConfigScopes["AVATAR"] = "AVATAR";
    DiscloudConfigScopes["BUILD"] = "BUILD";
    DiscloudConfigScopes["ID"] = "ID";
    DiscloudConfigScopes["MAIN"] = "MAIN";
    DiscloudConfigScopes["NAME"] = "NAME";
    DiscloudConfigScopes["RAM"] = "RAM";
    DiscloudConfigScopes["START"] = "START";
    DiscloudConfigScopes["TYPE"] = "TYPE";
    DiscloudConfigScopes["VERSION"] = "VERSION";
})(DiscloudConfigScopes || (exports.DiscloudConfigScopes = DiscloudConfigScopes = {}));
exports.discloudConfigRequiredScopes = {
    bot: [
        DiscloudConfigScopes.MAIN,
        DiscloudConfigScopes.NAME,
        DiscloudConfigScopes.TYPE,
        DiscloudConfigScopes.RAM,
        DiscloudConfigScopes.VERSION,
    ],
    site: [
        DiscloudConfigScopes.ID,
        DiscloudConfigScopes.MAIN,
        DiscloudConfigScopes.TYPE,
        DiscloudConfigScopes.RAM,
        DiscloudConfigScopes.VERSION,
    ],
    common: [
        DiscloudConfigScopes.MAIN,
        DiscloudConfigScopes.TYPE,
        DiscloudConfigScopes.RAM,
        DiscloudConfigScopes.VERSION,
    ],
};
exports.discloudConfigScopes = Object.values(DiscloudConfigScopes);
