// discloud.app/out/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.discloud = exports.User = exports.TeamAppStatus = exports.TeamAppManager = exports.TeamApp = exports.DiscloudApp = exports.DataManager = exports.CachedManager = exports.BaseTeamApp = exports.BaseManager = exports.BaseApp = exports.Base = exports.AppUploaded = exports.AppTeamManager = exports.AppTeam = exports.AppStatus = exports.AppManager = exports.AppBackup = exports.AppAptManager = exports.AppApt = exports.App = void 0;
const tslib_1 = require("tslib");
require("source-map-support/register");
const DiscloudApp_1 = tslib_1.__importDefault(require("./discloudApp/DiscloudApp"));
exports.DiscloudApp = DiscloudApp_1.default;
const AppAptManager_1 = tslib_1.__importDefault(require("./managers/AppAptManager"));
exports.AppAptManager = AppAptManager_1.default;
const AppManager_1 = tslib_1.__importDefault(require("./managers/AppManager"));
exports.AppManager = AppManager_1.default;
const AppTeamManager_1 = tslib_1.__importDefault(require("./managers/AppTeamManager"));
exports.AppTeamManager = AppTeamManager_1.default;
const BaseManager_1 = tslib_1.__importDefault(require("./managers/BaseManager"));
exports.BaseManager = BaseManager_1.default;
const CachedManager_1 = tslib_1.__importDefault(require("./managers/CachedManager"));
exports.CachedManager = CachedManager_1.default;
const DataManager_1 = tslib_1.__importDefault(require("./managers/DataManager"));
exports.DataManager = DataManager_1.default;
const TeamAppManager_1 = tslib_1.__importDefault(require("./managers/TeamAppManager"));
exports.TeamAppManager = TeamAppManager_1.default;
const App_1 = tslib_1.__importDefault(require("./structures/App"));
exports.App = App_1.default;
const AppApt_1 = tslib_1.__importDefault(require("./structures/AppApt"));
exports.AppApt = AppApt_1.default;
const AppBackup_1 = tslib_1.__importDefault(require("./structures/AppBackup"));
exports.AppBackup = AppBackup_1.default;
const AppStatus_1 = tslib_1.__importDefault(require("./structures/AppStatus"));
exports.AppStatus = AppStatus_1.default;
const AppTeam_1 = tslib_1.__importDefault(require("./structures/AppTeam"));
exports.AppTeam = AppTeam_1.default;
const AppUploaded_1 = tslib_1.__importDefault(require("./structures/AppUploaded"));
exports.AppUploaded = AppUploaded_1.default;
const Base_1 = tslib_1.__importDefault(require("./structures/Base"));
exports.Base = Base_1.default;
const BaseApp_1 = tslib_1.__importDefault(require("./structures/BaseApp"));
exports.BaseApp = BaseApp_1.default;
const BaseTeamApp_1 = tslib_1.__importDefault(require("./structures/BaseTeamApp"));
exports.BaseTeamApp = BaseTeamApp_1.default;
const TeamApp_1 = tslib_1.__importDefault(require("./structures/TeamApp"));
exports.TeamApp = TeamApp_1.default;
const TeamAppStatus_1 = tslib_1.__importDefault(require("./structures/TeamAppStatus"));
exports.TeamAppStatus = TeamAppStatus_1.default;
const User_1 = tslib_1.__importDefault(require("./structures/User"));
exports.User = User_1.default;
tslib_1.__exportStar(require("@discloudapp/api-types/v2"), exports);
tslib_1.__exportStar(require("@discloudapp/rest"), exports);
tslib_1.__exportStar(require("@discloudapp/util"), exports);
tslib_1.__exportStar(require("./@enum"), exports);
tslib_1.__exportStar(require("./@types"), exports);
tslib_1.__exportStar(require("./util"), exports);
/**
 * This is an instance of {@link DiscloudApp}
 *
 * Use this instead of instantiating the {@link DiscloudApp} class to use the library.
 *
 * @example
 *
 * ```js
 * // index.js
 * // Get instanced discloud
 * const { discloud } = require("discloud.app");
 *
 * // Set token
 * discloud.login("DISCLOUD_TOKEN");
 *
 * // Other file
 * // Get instanced discloud
 * const { discloud } = require("discloud.app");
 *
 * // ...
 *
 * async function () {
 *   await discloud.apps.fetch("ID"); // Promise<App>
 * }
 * ```
 */
exports.discloud = new DiscloudApp_1.default();
