// @discloudapp/util/out/GS.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GS = void 0;
const glob_1 = require("glob");
const os_1 = require("os");
const path_1 = require("path");
const IgnoreFiles_1 = require("./IgnoreFiles");
class GS {
    pattern;
    found = [];
    ignore;
    constructor(pattern, ignoreFileName, optionalIgnoreList = []) {
        this.pattern = pattern;
        this.ignore = new IgnoreFiles_1.IgnoreFiles({
            fileName: ignoreFileName,
            path: pattern,
            optionalIgnoreList,
        });
        if (Array.isArray(this.pattern)) {
            this.pattern = this.pattern.flatMap(pattern => [
                this.#normalizePath(pattern),
                (0, path_1.join)(this.#normalizePath(pattern), "**"),
            ]);
        }
        else {
            this.pattern = [
                this.#normalizePath(this.pattern),
                (0, path_1.join)(this.#normalizePath(this.pattern), "**"),
            ];
        }
        this.found = (0, glob_1.globSync)(this.pattern, {
            dot: true,
            ignore: this.ignore.list,
            windowsPathsNoEscape: (0, os_1.type)() === "Windows_NT",
        });
        if (this.found.includes(".")) {
            this.found.splice(this.found.indexOf("."), 1);
        }
        this.found = Array.from(new Set(this.found));
    }
    #normalizePath(path) {
        return (0, path_1.join)(...path.split(/[\\/]/));
    }
}
exports.GS = GS;
