// discloud.app/out/structures/App.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const AppApt_1 = tslib_1.__importDefault(require("./AppApt"));
const AppStatus_1 = tslib_1.__importDefault(require("./AppStatus"));
const AppTeam_1 = tslib_1.__importDefault(require("./AppTeam"));
const BaseApp_1 = tslib_1.__importDefault(require("./BaseApp"));
class App extends BaseApp_1.default {
    /**
     * Moderators IDs of your app
     */
    mods = new Set();
    constructor(discloudApp, data) {
        super(discloudApp, data);
        Object.defineProperties(this, {
            apt: { value: new AppApt_1.default(this.discloudApp, data) },
            status: { value: new AppStatus_1.default(this.discloudApp, data) },
            team: { value: new AppTeam_1.default(this.discloudApp, data) },
        });
        this._patch(data);
    }
    _patch(data) {
        if ("autoDeployGit" in data)
            this.autoDeployGit = data.autoDeployGit;
        if ("autoRestart" in data)
            this.autoRestart = data.autoRestart;
        if ("avatarURL" in data)
            this.avatarURL = data.avatarURL;
        if ("exitCode" in data)
            this.exitCode = data.exitCode;
        if ("lang" in data)
            this.lang = data.lang;
        if ("mainFile" in data)
            this.mainFile = data.mainFile;
        if ("mods" in data)
            if (Array.isArray(data.mods)) {
                this.mods.clear();
                for (const mod of data.mods) {
                    this.mods.add(mod);
                }
            }
        if ("name" in data)
            this.name = data.name;
        if ("online" in data)
            this.online = data.online;
        if ("ram" in data)
            this.ram = data.ram;
        if ("ramKilled" in data)
            this.ramKilled = data.ramKilled;
        // @ts-expect-error ts(2345) ts(2445)
        this.apt._patch(data);
        this.status._patch(data);
        this.team._patch(data);
        return super._patch(data);
    }
}
exports.default = App;
