// discloud.app/out/structures/AppBackup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fs_1 = require("fs");
const path_1 = require("path");
const process_1 = require("process");
const Base_1 = tslib_1.__importDefault(require("./Base"));
class AppBackup extends Base_1.default {
    /**
     * Your app id
     */
    appId;
    constructor(discloudApp, data) {
        super(discloudApp);
        this.appId = data.id;
        this._patch(data);
    }
    _patch(data) {
        this.url = data.url;
        if ("status" in data)
            this.status = data.status;
        return super._patch(data);
    }
    /**
     * Make backup of your application
     *
     * @param path - Backup path
     * @param fileName - Backup file name
     */
    async download(path = (0, process_1.cwd)(), fileName = this.appId) {
        if (!this.url)
            throw Error("Missing backup URL");
        path = path.toString();
        if (!(0, fs_1.existsSync)(path))
            (0, fs_1.mkdirSync)(path, { recursive: true });
        this.data = await fetch(this.url).then(res => res.arrayBuffer()).then(Buffer.from);
        const file = (0, path_1.join)(path, `${fileName}.${this.url.split(".").at(-1)}`);
        (0, fs_1.writeFileSync)(file, this.data);
        if ((0, fs_1.existsSync)(file))
            this.url = file;
        return this;
    }
}
exports.default = AppBackup;
