// discloud.app/out/structures/AppStatus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const util_1 = require("@discloudapp/util");
const Base_1 = tslib_1.__importDefault(require("./Base"));
class AppStatus extends Base_1.default {
    /**
     * Your app id
     */
    appId;
    constructor(discloudApp, data) {
        super(discloudApp);
        this.appId = data.id;
    }
    _patch(data) {
        if ("container" in data)
            this.container = data.container;
        if ("cpu" in data)
            this.cpu = data.cpu;
        if ("last_restart" in data)
            this.lastRestart = data.last_restart;
        if ("memory" in data) {
            this.memory = data.memory;
            const matched = data.memory?.match(/[\d.]+/g) ?? [];
            this.memoryUsage = (0, util_1.calculatePercentage)(matched[0], matched[1]);
        }
        if ("netIO" in data)
            this.netIO = data.netIO;
        if ("ssd" in data)
            this.ssd = data.ssd;
        if ("startedAt" in data) {
            this.startedAt = new Date(data.startedAt);
            this.startedAtTimestamp = this.startedAt.valueOf();
        }
        return super._patch(data);
    }
    get app() {
        return this.discloudApp.apps.cache.get(this.appId);
    }
    fetch() {
        return this.discloudApp.apps.status(this.appId);
    }
}
exports.default = AppStatus;
