// discloud.app/out/structures/AppUploaded.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Base_1 = tslib_1.__importDefault(require("./Base"));
class AppUploaded extends Base_1.default {
    /**
     * Your app id
     */
    appId;
    constructor(discloudApp, data) {
        super(discloudApp);
        this.appId = data.id;
        this._patch(data);
    }
    _patch(data) {
        if ("addedAtTimestamp" in data)
            this.addedAtTimestamp = data.addedAtTimestamp;
        if ("avatarURL" in data)
            this.avatarURL = data.avatarURL;
        if ("autoRestart" in data)
            this.autoRestart = data.autoRestart;
        if ("lang" in data)
            this.lang = data.lang;
        if ("mainFile" in data)
            this.mainFile = data.mainFile;
        if ("name" in data)
            this.name = data.name;
        if ("ram" in data)
            this.ram = data.ram;
        if ("type" in data)
            this.type = data.type;
        if ("version" in data)
            this.version = data.version;
        return super._patch(data);
    }
    get app() {
        return this.discloudApp.apps.cache.get(this.appId);
    }
}
exports.default = AppUploaded;
