// @discloudapp/rest/out/utils/contants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultRestOptions = exports.DefaultUserAgent = exports.cpu_arch = exports.os_platform = exports.os_release = exports.os_name = void 0;
const v2_1 = require("@discloudapp/api-types/v2");
const fs_1 = require("fs");
const os_1 = require("os");
const path_1 = require("path");
exports.os_name = (0, os_1.type)();
exports.os_release = (0, os_1.release)().split?.(".").slice(0, 2).join(".") ?? (0, os_1.release)();
exports.os_platform = (0, os_1.platform)();
exports.cpu_arch = (0, os_1.arch)();
const DiscloudAppPackageLocation = (0, path_1.join)(__dirname, "..", "..", "..", "..", "discloud.app", "package.json");
let discloudAppVersion = "";
if ((0, fs_1.existsSync)(DiscloudAppPackageLocation))
    discloudAppVersion = `/${JSON.parse((0, fs_1.readFileSync)(DiscloudAppPackageLocation, "utf8")).version}`;
exports.DefaultUserAgent = `DiscloudApp${discloudAppVersion} (${exports.os_name} ${exports.os_release}; ${exports.os_platform}; ${exports.cpu_arch})`;
exports.DefaultRestOptions = {
    api: "https://api.discloud.app",
    globalRequestsPerMinute: 60,
    headers: {
        "User-Agent": exports.DefaultUserAgent,
    },
    version: v2_1.APIVersion,
};
