// discloud.app/out/managers/AppTeamManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const v2_1 = require("@discloudapp/api-types/v2");
const util_1 = require("@discloudapp/util");
const zod_1 = tslib_1.__importDefault(require("zod"));
const BaseManager_1 = tslib_1.__importDefault(require("./BaseManager"));
/**
 * Manager for Team on your application on Discloud
 */
class AppTeamManager extends BaseManager_1.default {
    constructor(discloudApp) {
        super(discloudApp);
    }
    /**
     * Add a mod for you application on Discloud
     *
     * @param appID - The app id
     * @param modID - The mod id
     * @param perms - The permissions for the mod. See {@link ModPermissionsResolvable}
     */
    async create(appID, modID, perms) {
        zod_1.default.string().parse(appID);
        zod_1.default.string().parse(modID);
        const data = await this.discloudApp.rest.post(v2_1.Routes.appTeam(appID), {
            body: {
                modID,
                perms: new util_1.ModPermissionsBF(perms).toArray(),
            },
        });
        return data.app;
    }
    /**
     * Edit permissions of a mod for you application on Discloud
     *
     * @param appID - The app id
     * @param modID - The mod id
     * @param perms - The permissions for the mod. See {@link ModPermissionsResolvable}
     */
    async edit(appID, modID, perms) {
        zod_1.default.string().parse(appID);
        zod_1.default.string().parse(modID);
        const data = await this.discloudApp.rest.put(v2_1.Routes.appTeam(appID), {
            body: {
                modID,
                perms: new util_1.ModPermissionsBF(perms).toArray(),
            },
        });
        return data.app;
    }
    /**
     * Remove a mod from you application on Discloud
     *
     * @param appID - The app id
     * @param modID - The mod id
     */
    async delete(appID, modID) {
        zod_1.default.string().parse(appID);
        zod_1.default.string().parse(modID);
        const data = await this.discloudApp.rest.delete(v2_1.Routes.appTeam(appID, modID));
        return data;
    }
    /**
     * Get mods information for your app on Discloud
     */
    async fetch(appID) {
        zod_1.default.string().parse(appID);
        const data = await this.discloudApp.rest.get(v2_1.Routes.appTeam(appID));
        return data.team;
    }
}
exports.default = AppTeamManager;
