// @discloudapp/api-types/rest/v2/routes/App.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.App = void 0;
class App {
    /**
     * - GET - `/app/{appID}`
     */
    static app(appID) {
        return `/app/${appID}`;
    }
    /**
     * - PUT - `/app/{appID}/apt`
     * - DELETE - `/app/{appID}/apt`
     */
    static apt(appID) {
        return `/app/${appID}/apt`;
    }
    /**
     * - GET - `/app/{appID}/backup`
     */
    static backup(appID) {
        return `/app/${appID}/backup`;
    }
    /**
     * - PUT - `/app/{appID}/commit`
     */
    static commit(appID) {
        return `/app/${appID}/commit`;
    }
    /**
     * - PUT - `/app/{appID}/console`
     */
    static console(appID) {
        return `/app/${appID}/console`;
    }
    /**
     * - DELETE - `/app/{appID}/delete`
     */
    static delete(appID) {
        return `/app/${appID}/delete`;
    }
    /**
     * - GET - `/app/{appID}/logs`
     */
    static logs(appID) {
        return `/app/${appID}/logs`;
    }
    /**
     * - PUT - `/app/{appID}/profile`
     */
    static profile(appID) {
        return `/app/${appID}/profile`;
    }
    /**
     * - PUT - `/app/{appID}/ram`
     */
    static ram(appID) {
        return `/app/${appID}/ram`;
    }
    /**
     * - PUT - `/app/{appID}/restart`
     */
    static restart(appID) {
        return `/app/${appID}/restart`;
    }
    /**
     * - PUT - `/app/{appID}/start`
     */
    static start(appID) {
        return `/app/${appID}/start`;
    }
    /**
     * - GET - `/app/{appID}/status`
     */
    static status(appID) {
        return `/app/${appID}/status`;
    }
    /**
     * - PUT - `/app/{appID}/stop`
     */
    static stop(appID) {
        return `/app/${appID}/stop`;
    }
    static team(appID, modID) {
        return modID ? `/app/${appID}/team/${modID}` : `/app/${appID}/team`;
    }
}
exports.App = App;
