// @discloudapp/api-types/rest/v2/routes/Team.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Team = void 0;
class Team {
    /**
     * - GET - `/team`
     */
    static team() {
        return "/team";
    }
    /**
     * - GET - `/app/{appID}/backup`
     */
    static backup(appID) {
        return `/team/${appID}/backup`;
    }
    /**
     * - PUT - `/app/{appID}/commit`
     */
    static commit(appID) {
        return `/team/${appID}/commit`;
    }
    /**
     * - GET - `/team/{appID}/logs`
     */
    static logs(appID) {
        return `/team/${appID}/logs`;
    }
    /**
     * - PUT - `/team/{appID}/ram`
     */
    static ram(appID) {
        return `/team/${appID}/ram`;
    }
    /**
     * - PUT - `/app/{appID}/restart`
     */
    static restart(appID) {
        return `/team/${appID}/restart`;
    }
    /**
     * - PUT - `/app/{appID}/start`
     */
    static start(appID) {
        return `/team/${appID}/start`;
    }
    /**
     * - PUT - `/app/{appID}/status`
     */
    static status(appID) {
        return `/team/${appID}/status`;
    }
    /**
     * - PUT - `/app/{appID}/stop`
     */
    static stop(appID) {
        return `/team/${appID}/stop`;
    }
}
exports.Team = Team;
